### DISCUSSION TITLE
ABP why reset frame counter?

### DISCUSSION THREAD

**Discussion by Tom1278** [Lora-net/LoRaMac-node#1053] [Q&A] [answered]
I'm trying out activation by personalization but the first thing the LoRa stack does is reset the frame counter by calling this function: ResetFCnts
I will be implementing my own non-volatile memory that will store the frame counter and I'm looking for a place in the LoRa stack that I call during the startup to set the frame counter.
Of course I could make my own, for example: CryptoCtx.NvmCtx->FCntList.FCntUp = x;
But I would rather do the correct way if there is one and maybe someone can help me out and point me in the right direction?
Many thanks!

**Comment by mluis1** [comment, marked as answer]
At stack initialization everything is reset to defaults and this is normal.

After the stack initialization one must restore the previous context from a non volatile memory.
As you can see in below code excerpt the stack is initialized by calling `LoRaMacInitialization` and then the NVM context is restored by calling `NvmDataMgmtRestore`.
It is to be noted that in case the restore function didn't succeed then the stack behaves like if it was the first time it was run.

```c
LmHandlerErrorStatus_t LmHandlerInit( LmHandlerCallbacks_t *handlerCallbacks,
                                      LmHandlerParams_t *handlerParams )
{
    //
    uint16_t nbNvmData = 0;
    MibRequestConfirm_t mibReq;
    LmHandlerParams = handlerParams;
    LmHandlerCallbacks = handlerCallbacks;

    LoRaMacPrimitives.MacMcpsConfirm = McpsConfirm;
    LoRaMacPrimitives.MacMcpsIndication = McpsIndication;
    LoRaMacPrimitives.MacMlmeConfirm = MlmeConfirm;
    LoRaMacPrimitives.MacMlmeIndication = MlmeIndication;
    LoRaMacCallbacks.GetBatteryLevel = LmHandlerCallbacks->GetBatteryLevel;
    LoRaMacCallbacks.GetTemperatureLevel = LmHandlerCallbacks->GetTemperature;
    LoRaMacCallbacks.NvmDataChange  = NvmDataMgmtEvent;
    LoRaMacCallbacks.MacProcessNotify = LmHandlerCallbacks->OnMacProcess;

    IsClassBSwitchPending = false;
    IsUplinkTxPending = false;

    if( LoRaMacInitialization( &LoRaMacPrimitives, &LoRaMacCallbacks, LmHandlerParams->Region ) != LORAMAC_STATUS_OK )
    {
        return LORAMAC_HANDLER_ERROR;
    }

    // Restore data if required
    nbNvmData = NvmDataMgmtRestore( );

    // Try to restore from NVM and query the mac if possible.
    if( nbNvmData > 0 )
    {
        LmHandlerCallbacks->OnNvmDataChange( LORAMAC_HANDLER_NVM_RESTORE, nbNvmData );
    }
    else
    {
...
}
```

In the future it would be nice if you could post this kind of questions on the project [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab. It is a better place to engage discussions and then we can agree if it is an issue or not.
If you don't mind I move this issue to the [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab.

**Reply by Tom1278** [reply]
Thank you!
